### PR TITLE
Short form bookmark url support

### PR DESCRIPTION
**Pull request by mortenbruhn**
## Description
- Added support for using c:geo to open coord.info short urls for bookmarks

## Related issues

None known

## Additional context
It has not (yet) been spun up, not on virtual or a phone as APK.

Comments and suggestions welcome.

### CODE REVIEW

**Comment by bekuno** [comment]
Ok to test

**Comment by bekuno** [comment]
Retest this please

**Comment by eddiemuc** [comment]
> > Retest this please
> 
> I am guessing this is not directed at me?

It is not :-) 
"retest this please" is a command to the underlying CI to rerun compilation and unit-testing of the code.

**Comment by mortenbruhn** [comment]
> > > Retest this please
> > 
> > 
> > I am guessing this is not directed at me?
> 
> It is not :-) "retest this please" is a command to the underlying CI to rerun compilation and unit-testing of the code.

soooo.. what happens next ?

**Comment by bekuno** [comment]
> > It is not :-) "retest this please" is a command to the underlying CI to rerun compilation and unit-testing of the code.
> 
> soooo.. what happens next ?

You should see the success state of the CI test here, isn´t it?

grafik

The `Details` link goes to the CI Build overview for this PR.

Please have a look at the **Lint issues** result.
There is a new warning from your build.
Please resolve the issue. I believe that the `toUpperCase(Locale.US)` should be fine.

And someone from the team will probaly check your PR.
I can not say, who has time and knowlege to do that.

File: main/src/main/java/cgeo/geocaching/connector/gc/ImportBookmarkLinks.java [hunk 1]
```diff
@@ -9,13 +9,90 @@
...
+    public static final String BOOKMARK_LIST_API_PREFIX =
+            "https://www.geocaching.com/plan/api/gpx/list/";
+
+    interface UrlToIdParser {
+        @Nullable String tryExtractFromIntentUrl(@Nullable String intentUrl);
+    }
+
+    static class SingleUrlToIdParser implements UrlToIdParser {
+        private final Pattern matcherPattern;
+        // Android API 21 apparently does not support named capturing groups in Pattern
```
Comment (eddiemuc) [inline main/src/main/java/cgeo/geocaching/connector/gc/ImportBookmarkLinks.java:33]: Thats correct (and very sad), unfortunately we can't use group capturing in c:geo....

What I usually do however is put the group names into the regex anyway (for better readability of the pattern) and comment somewhere which group index corresponds to which group. If at some day we drop support for SDK21 we will be able to use named capturing group.

File: main/src/main/java/cgeo/geocaching/connector/gc/ImportBookmarkLinks.java [hunk 2]
```diff
@@ -9,13 +9,90 @@
...
+import javax.annotation.RegEx;
+
 public class ImportBookmarkLinks extends AppCompatActivity {
 
+    private final UrlToIdParser intentUrlParser = defaultBookmarkListUrlToIdParser();
+
+    public static final String BOOKMARK_LIST_API_PREFIX =
+            "https://www.geocaching.com/plan/api/gpx/list/";
+
+    interface UrlToIdParser {
```
Comment (eddiemuc) [inline main/src/main/java/cgeo/geocaching/connector/gc/ImportBookmarkLinks.java:27]: No need to make up an interface for this, you may use the generic functional interface `cgeo.geocaching.utils.functions.Func1` 

(NB: I just noted that apparently my IDE does not complain the usage of the java.util.function interfaces any more like it used to. Has something changed here? I will investigate, maybe we can finally get rid of cgeo.geocaching.utils.functions.... :-))

File: main/src/main/java/cgeo/geocaching/connector/gc/ImportBookmarkLinks.java [hunk 3]
```diff
@@ -9,13 +9,90 @@
...
+                }
+            }
+            return null;
+        }
+    }
+
+    static UrlToIdParser defaultBookmarkListUrlToIdParser() {
+        return new AggregatedUrlToIdParser(
+                new SingleUrlToIdParser(
+                        "^https?://(?:www\\.)?geocaching\\.com/plan/lists/([A-Z0-9]+)",
```
Comment (eddiemuc) [inline main/src/main/java/cgeo/geocaching/connector/gc/ImportBookmarkLinks.java:86]: As stated above: it might make sense to use naming groups in the regex anyway and comment which group index corresponds to the naming group

File: main/src/main/java/cgeo/geocaching/connector/gc/ImportBookmarkLinks.java [hunk 4]
```diff
@@ -9,13 +9,90 @@
...
+                @RegEx @NonNull final String regex,
+                final int groupToExtract
+        ) {
+            this.matcherPattern = Pattern.compile(regex, Pattern.CASE_INSENSITIVE);
+            this.groupToExtract = groupToExtract;
+        }
+
+        @Override
+        public String tryExtractFromIntentUrl(final String intentUrl) {
+            if (intentUrl == null) {
```
Comment (eddiemuc) [inline main/src/main/java/cgeo/geocaching/connector/gc/ImportBookmarkLinks.java:47]: Instead of the logic in the method, use a call to `TextUtils.getMatch(@Nullable final String data, final Pattern pattern, final boolean trim, final int group, @Nullable final String defaultValue, final boolean last)`

File: main/src/test/java/cgeo/geocaching/connector/gc/ImportBookmarkLinksTest.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
+package cgeo.geocaching.connector.gc;
+
+import org.junit.Test;
+import static org.assertj.core.api.Java6Assertions.assertThat;
+
+public class ImportBookmarkLinksTest {
```
Comment (eddiemuc) [inline main/src/test/java/cgeo/geocaching/connector/gc/ImportBookmarkLinksTest.java:6]: Kudos and thumbs up for adding a test class!

**Review by eddiemuc — APPROVED** [review]
Several comments, but nothing cricital.

**Comment by eddiemuc** [comment]
@mortenbruhn Thanks a lot for your contribution! I reviewed the PR and found only non-critical things. Special kudos for adding unit tests, really appreciated! 
One question though: which issue does this PR refer to? If there's none you kind of "skipped" the normal discussion process where we decide which features we would like to have in c:geo and which not.

@moving-bits importing bookmarks seems to be your iriginal work, would you want to review as well?

**Comment by mortenbruhn** [comment]
> > > It is not :-) "retest this please" is a command to the underlying CI to rerun compilation and unit-testing of the code.
> > 
> > 
> > soooo.. what happens next ?
> 
> You should see the success state of the CI test here, isn´t it?

Yes. Seemed happy?

> grafik
> 
> The `Details` link goes to the CI Build overview for this PR.
> 
> Please have a look at the **Lint issues** result. There is a new warning from your build. Please resolve the issue. I believe that the `toUpperCase(Locale.US)` should be fine.

Done. `Locale.getDefault()` seems more appropriate IMO.
 
> And someone from the team will probaly check your PR. I can not say, who has time and knowlege to do that.

Great. Looking forward to it :)

File: main/src/main/java/cgeo/geocaching/connector/gc/ImportBookmarkLinks.java (hunk 2 above)
Comment (mortenbruhn) [reply main/src/main/java/cgeo/geocaching/connector/gc/ImportBookmarkLinks.java]: I still get the warning:
`Class requires API level 24 (current min is 21): java.util.function.Function`

Also: I made the interface to be domain-specific. It signals a specific responsibility. I have, however, now made it extend `Func1`

File: main/src/main/java/cgeo/geocaching/connector/gc/ImportBookmarkLinks.java (hunk 3 above)
Comment (mortenbruhn) [reply main/src/main/java/cgeo/geocaching/connector/gc/ImportBookmarkLinks.java]: Done

File: main/src/test/java/cgeo/geocaching/connector/gc/ImportBookmarkLinksTest.java (hunk 5 above)
Comment (mortenbruhn) [reply main/src/test/java/cgeo/geocaching/connector/gc/ImportBookmarkLinksTest.java:6]: Thank you :)

File: main/src/main/java/cgeo/geocaching/connector/gc/ImportBookmarkLinks.java (hunk 1 above)
Comment (mortenbruhn) [reply main/src/main/java/cgeo/geocaching/connector/gc/ImportBookmarkLinks.java:40]: Done

**Comment by mortenbruhn** [comment]
> @mortenbruhn Thanks a lot for your contribution! I reviewed the PR and found only non-critical things. Special kudos for adding unit tests, really appreciated! One question though: which issue does this PR refer to? If there's none you kind of "skipped" the normal discussion process where we decide which features we would like to have in c:geo and which not.

I did. It seemed like an oversight to support opening links from both geocaching.com and coord.info for both geocache listings and trackables, but only geocaching.com for bookmark lists.

File: main/AndroidManifest.xml [hunk 6]
```diff
@@ -705,6 +705,20 @@
...
             android:exported="true">
 
+            <!-- geocaching.com URL via coord.info redirection -->
+            <intent-filter>
+                <action android:name="android.intent.action.VIEW" />
+
+                <category android:name="android.intent.category.DEFAULT" />
+                <category android:name="android.intent.category.BROWSABLE" />
+
+                <data android:host="coord.info" />
```
Comment (moving-bits) [inline main/AndroidManifest.xml:715]: Thanks @mortenbruhn for your contribution, very welcomed. 

From a quick test in the emulator it seems a small addition to this code block in `AndroidManifest.xml` is required to make it work:

```
<data android:scheme="http" />
<data android:scheme="https" />
```

Without those lines, a link like `https://www.coord.info/BM/...` in an email did not open in c:geo but in the web browser, whereas with those two lines it opens in c:geo. Tested in the emulator with both GMail and K9 as mail client.

Comment (mortenbruhn) [reply main/AndroidManifest.xml:717]: Right you are. Done :+1:

**Comment by moving-bits** [comment]
Thanks for the quick addition. LGTM, so let's put it in the nightlies. Thanks @mortenbruhn for your contribution!
